pymc3/step_methods/hmc/nuts.py

```diff
@@ -239,40 +241,46 @@ def extend(self, direction):
             self.right = tree.right
         else:
             tree, diverging, turning = self._build_subtree(
-                self.left, self.depth, floatX(np.asarray(- self.step_size)))
+                self.left, self.depth, floatX(np.asarray(-self.step_size)))
             self.left = tree.right
 
-        ok = not (diverging or turning)
-        if ok and bern(min(1, tree.size / self.size)):
+        self.depth += 1
+        self.accept_sum += tree.accept_sum
+        self.n_proposals += tree.n_proposals
+
+        if diverging or turning:
+            return diverging, turning
+
+        size1, size2 = self.log_size, tree.log_size
+        if logbern(size2 - size1):
             self.proposal = tree.proposal
 
-        self.depth += 1
-        self.size += tree.size
-        # TODO why not +=
-        #self.accept_sum += tree.accept_sum
-        self.accept_sum = tree.accept_sum
-        #self.n_proposals += tree.n_proposals
-        self.n_proposals = tree.n_proposals
+        self.log_size = np.logaddexp(self.log_size, tree.log_size)
+        self.p_sum[:] += tree.p_sum
 
         left, right = self.left, self.right
-        span = right.q - left.q
-        turning = turning or (span.dot(left.p) < 0) or (span.dot(right.p) < 0)
+        p_sum = self.p_sum
+        turning = (p_sum.dot(left.v) <= 0) or (p_sum.dot(right.v) <= 0)
```

Comment:
This is different. The old implementation is actually somewhat wrong: It does not take into account the metric, but assumes a standard euclidean space. This was an assumption in the original paper, but in general we need to use the scalar product defined by the inverse mass matrix. (see [the old stan version](https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/hmc/nuts_classic/dense_e_nuts_classic.hpp#L27)) Stan switched to a new version recently (I linked a paper with a description of that somewhere in the comments – Ctrl-F Betancourt). As far as I understand that was mostly to make riemannian hmc easier.